[ST] - Add support of provisioning Kind cluster

see-quick opened the pull request:
### Type of change

- Enhancement / new feature
- Refactoring
- Documentation

### Description

Currently, a `minikube` does not have support for IPv6 and thus, it is impossible to test such configuration on Azure pipelines using our current setup (i.e., `minikube`). Fortunately, the [kind](https://kind.sigs.k8s.io/) could solve it because it supports IPv6 or dual stack. 

I have modified our `setup-kubernetes.sh` script to be able also provision the `kind` cluster. In the future, we can also use this as a part of testing farm infrastructure where we can deploy multi-node clusters (because there we have unlimited resources). To contrast in Azure Pipelines we are limited to small machines (i.e., 6GB RAM and 2 CPU) and here we will stick with single-node I assume but (we could easily test dual and IPv6 configurations). 

### Checklist

(IPv4 profiles)
- [x] Check smoke profile 
- [x] Check acceptance profile 
- [x] Make sure all tests pass

Frawless commented on .azure/scripts/setup-kubernetes.sh:
```diff
@@ -1,20 +1,25 @@
-DEFAULT_MINIKUBE_CPU=$(awk '$1~/cpu[0-9]/{usage=($2+$4)*100/($2+$4+$5); print $1": "usage"%"}' /proc/stat | wc -l)
+DEFAULT_KIND_MEMORY=$(free -m | grep "Mem" | awk '{print $2}')
+DEFAULT_KIND_CPU=$(awk '$1~/cpu[0-9]/{usage=($2+$4)*100/($2+$4+$5); print $1": "usage"%"}' /proc/stat | wc -l)
 
-MINIKUBE_MEMORY=${MINIKUBE_MEMORY:-$DEFAULT_MINIKUBE_MEMORY}
-MINIKUBE_CPU=${MINIKUBE_CPU:-$DEFAULT_MINIKUBE_CPU}
+KIND_MEMORY=${KIND_MEMORY:-$DEFAULT_KIND_MEMORY}
+KIND_CPU=${KIND_CPU:-$DEFAULT_KIND_CPU}
```
I think you shouldn't rename it like that. These options should be common for all clusters available via the script.

see-quick replied:
Yeah, it's just left-over I would name it f.e. `DEFAULT_CLUSTER_MEMORY` etc.

Frawless commented on .azure/scripts/setup-kubernetes.sh:
```diff
@@ -78,47 +143,85 @@ if [ "$TEST_CLUSTER" = "minikube" ]; then
-      set +ex
+elif [ "$TEST_CLUSTER" = "kind" ]; then
+    install_kubectl
+    install_kubernetes_provisioner
 
-      docker exec "minikube" bash -c "echo '$(cat $HOME/.docker/config.json)'| sudo tee -a /var/lib/kubelet/config.json > /dev/null && sudo systemctl restart kubelet"
+    setup_kube_directory
```
Wouldn't be better to run the common parts before the if statement to avoid redundant lines?

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels acceptance

see-quick commented on .packit.yaml:
```diff
@@ -140,9 +140,45 @@ jobs:
           name: sanity
+
+  ###############################################################################################
+  ##################################    IPv6/dual jobs   ########################################
```
I wonder if we should merge one acceptance job, which will trigger also IPv6 and dual configuration instead of only IPv4

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

im-konge commented on .azure/scripts/setup-kubernetes.sh:
```diff
@@ -78,47 +142,108 @@ if [ "$TEST_CLUSTER" = "minikube" ]; then
+    add_docker_hub_credentials_to_kubernetes "$TEST_CLUSTER"
+    setup_registry_proxy
+
+elif [ "$TEST_CLUSTER" = "kind" ]; then
```
Would it make sense to have two separate scripts for `kind` and `minikube` setup, to make this script more clear? I guess that it can be confusing having both in same script file.

see-quick replied:
Well, there is a lot of shared code between those two installations. So I wanted to have it as one script. I could make a `common.sh` and divide these 2 scripts, but I don't know how much value it gives. But if you really want it I can divide it :)

see-quick replied:
I have divided them into two separate scripts ;)

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:
```diff
@@ -471,6 +473,9 @@ void testBuildOtherPluginTypeWithAndWithoutFileName(ExtensionContext extensionCo
     @Tag(ACCEPTANCE)
     @ParallelTest
     void testBuildPluginUsingMavenCoordinatesArtifacts(ExtensionContext extensionContext) {
+        // using kind we encounter (error building image: deleting file system after stage 0: unlinkat //product_uuid: device or resource busy)
+        assumeFalse(KubeClusterResource.getInstance().isKind());
+
```
Maybe create annotation and skip the whole suite? I guess it will help you also for `regression` pipelines

see-quick replied:
> Maybe create annotation and skip the whole suite? I guess it will help you also for regression pipelines

I don't know if all KafkaConnectBuilder test cases will fail. I am not fully sure but on Thursday I execute whole suite and only 2-3 test cases failed. So I can for sure create that annotation but I will skip only those, which will fail.

see-quick commented:
/packit test --labels acceptance

scholzj reviewed (commented):
Do you plan to update all the other repositories as well? If not, it might be worth to keep the original names etc. to make it easier to copy between the CI configurations.

see-quick commented:
> Do you plan to update all the other repositories as well? If not, it might be worth to keep the original names etc. to make it easier to copy between the CI configurations.

Yeah, but first I want to make it stable in the operator repo. In this PR I have added only acceptance profile but in the next I would also add regression and I think there will be some failures related to IPv6 (I hope not :D)

see-quick commented:
/packit test --labels acceptance

scholzj commented:
> > Do you plan to update all the other repositories as well? If not, it might be worth to keep the original names etc. to make it easier to copy between the CI configurations.
> 
> Yeah, but first I want to make it stable in the operator repo. In this PR I have added only acceptance profile but in the next I would also add regression and I think there will be some failures related to IPv6 (I hope not :D)

So, if there are no plans for it right now, maybe you should not change the things in `./azure/templates`. They do not seem to add any value and they seem to just complicate the live when sharing the parts of the CI.

see-quick commented:
> So, if there are no plans for it right now, maybe you should not change the things in ./azure/templates. They do not seem to add any value and they seem to just complicate the live when sharing the parts of the CI.



> > > Do you plan to update all the other repositories as well? If not, it might be worth to keep the original names etc. to make it easier to copy between the CI configurations.
> > 
> > 
> > Yeah, but first I want to make it stable in the operator repo. In this PR I have added only acceptance profile but in the next I would also add regression and I think there will be some failures related to IPv6 (I hope not :D)
> 
> So, if there are no plans for it right now, maybe you should not change the things in `./azure/templates`. They do not seem to add any value and they seem to just complicate the live when sharing the parts of the CI.



> > > Do you plan to update all the other repositories as well? If not, it might be worth to keep the original names etc. to make it easier to copy between the CI configurations.
> > 
> > 
> > Yeah, but first I want to make it stable in the operator repo. In this PR I have added only acceptance profile but in the next I would also add regression and I think there will be some failures related to IPv6 (I hope not :D)
> 
> So, if there are no plans for it right now, maybe you should not change the things in `./azure/templates`. They do not seem to add any value and they seem to just complicate the live when sharing the parts of the CI.

Yeah, I will undo such changes :).

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels acceptance

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels smoke

see-quick commented:
/packit test --labels acceptance

see-quick commented:
Specific fixes to regression on Testing Farm and IPv6-related stuff I will continue on different PR. The main aim of this PR was achieved and we can easily now test acceptance + smoke pipelines (without ZK issues).

see-quick commented:
/packit test --labels smoke

scholzj commented on systemtest/src/main/java/io/strimzi/systemtest/annotations/KindNotSupport.java:
```diff
@@ -0,0 +1,35 @@
+@Target({ElementType.METHOD, ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@ExtendWith(KindNotSupportCondition.class)
+public @interface KindNotSupport {
```
No strong opinions ... but should be be `KindNotSupported`? Or `NoKindSupport`? `KindNotSupport` sounds super weird. (but not a blocker from my point of view)

see-quick replied:
Yeah, `KindNotSupported` sounds better thanks for it :)

scholzj commented on systemtest/src/main/java/io/strimzi/systemtest/Environment.java:
```diff
@@ -287,13 +295,29 @@ public static boolean useLatestReleasedBridge() {
+            } catch (UnknownHostException e) {
+                throw new RuntimeException(e);
+            }
+            LOGGER.info("Using hostname:{}", hostname);
```
Is that ome left over debug message? If not, should i tbe at least properly formatted and have some meaning?

```suggestion
            LOGGER.info("Using container registry hostname: {}", hostname);
```

see-quick replied:
Updated :) thanks

see-quick commented:
/packit test --labels smoke

im-konge commented on .azure/scripts/setup-kind.sh:
```diff
@@ -0,0 +1,200 @@
+
+function install_kubernetes_provisioner {
+
+    if [ "$TEST_CLUSTER" = "kind" ]; then
```
It will be always `kind` when using this script, right?

im-konge commented on .azure/scripts/setup-kind.sh:
```diff
@@ -0,0 +1,200 @@
+install_kubectl
+install_kubernetes_provisioner
+
+if [ "$TEST_CLUSTER" = "kind" ]; then
```
Here as well

im-konge reviewed (approved):
Just two nits, thanks for the PR :) good job

see-quick commented:
/packit test --labels smoke
